File: Design-Patterns-Using-Kotlin/Proxy/ProxyLazy.kt
interface Subject {
    fun request()
}

class RealSubject : Subject {
    override fun request() {
        println("Concrete Subject Request Method")
    }
}

class Proxy : Subject {
    private var realSubject: RealSubject? = null

    override fun request() {
        if (realSubject == null) {
            realSubject = RealSubject() // Lazy Initialization
        }
        realSubject!!.request()
    }
}

// Client Code
fun main() {
    val proxy = Proxy()
    proxy.request()
}

/*
Concrete Subject Request Method
*/

File: Design-Patterns-Using-Swift/Proxy/ProxyLazy.swift
import Foundation

// Subject protocol
protocol Subject {
    func request()
}

// RealSubject class conforming to Subject protocol
class RealSubject: Subject {
    func request() {
        print("Concrete Subject Request Method")
    }
}

// Proxy class conforming to Subject protocol
class Proxy: Subject {
    private var realSubject: RealSubject?

    func request() {
        if realSubject == nil {
            realSubject = RealSubject() // Lazy Initialization
        }
        realSubject?.request()
    }
}

// Client Code
let proxy = Proxy()
proxy.request()

/*
Concrete Subject Request Method
*/
